packages/serverpod/lib/src/util/service_locator/service_locator.dart

```diff
@@ -0,0 +1,119 @@
+import 'locator_exceptions.dart';
+
+/// A service locator interface that allows for locating services by type or key.
+// This interface is used to decouple service registration and retrieval.
+abstract interface class ServiceLocator {
+  /// Locates a service by its type.
+  /// Throws [ServiceNotFoundException] if the service is not found.
+  /// @param T The type of the service to locate.
+  /// @return The service of type [T] if found.
+  T locateType<T>();
```

Comment:
Have you considered combining the look-up functions? I think that might read a bit nicer when used.

```suggestion
  T locate<T>([String? key]);
```

or

```suggestion
  T locate<T>({String? key});
```

Reply:
I've consolidated the methods as suggested and updated the tests